File: bench/kernels/bench_collatz.js
// bench_collatz.js - total Collatz steps for n=1..1,000,000 (qjs)
function collatzSteps(n) {
    let steps = 0;
    while (n !== 1) {
        n = (n % 2) ? n * 3 + 1 : n / 2;
        steps++;
    }
    return steps;
}

const N = 1000000;
const t0 = Date.now();
let total = 0;
for (let i = 1; i <= N; i++)
    total += collatzSteps(i);
const elapsed = Date.now() - t0;
print("result: " + total);
print("elapsed: " + elapsed + " ms");

File: bench/kernels/bench_collatz.rs
// bench_collatz.rs - total Collatz steps for n=1..1,000,000
use std::time::Instant;

fn collatz_steps(mut n: i64) -> i64 {
    let mut steps = 0i64;
    while n != 1 {
        n = if n & 1 != 0 { n * 3 + 1 } else { n >> 1 };
        steps += 1;
    }
    steps
}

fn main() {
    let n = 1_000_000i64;
    let t0 = Instant::now();
    let mut total = 0i64;
    for i in 1..=n {
        total += collatz_steps(i);
    }
    let elapsed = t0.elapsed().as_millis();
    println!("result: {}", total);
    println!("elapsed: {} ms", elapsed);
}
